File: func-pipes/static/js/power2/nodes/heater.js
/*
  nodes/heater.js — Electric Heater
  ─────────────────────────────────────────────────────────────────────────────
  Extends Load with thermal simulation and dynamic power draw. As the heater
  warms up its element draws progressively more power (up to the rated watts).
  A built-in thermostat (heatSwitch) trips the element off at maxTemp and
  resets it once the unit cools to resetTemp. A minimum standby draw
  (minWatts) is always present even when the element is off.

  This class demonstrates how a developer adds a custom node type:
    1. Extend a suitable base (Load, in this case).
    2. Declare static type, label, group, catalog.
    3. Override defaults() to add your extra fields.
    4. Override apply() to customise the power-draw amount.
    5. Override tick() to add per-frame behaviour (thermal update + thermostat).
    6. Override reset() to clear your extra state.
    7. Call NodeRegistry.register(Heater) to make it available in the graph.

  Extra state
  ───────────
  temperature   number  — simulated temperature in °C (0–maxTemp range)
  heatState     string  — 'cold' | 'warming' | 'hot'
  heatSwitch    bool    — thermostat: true = element on, false = tripped off
  currentWatts  number  — live draw (minWatts … watts), updated each tick
  heatRate      number  — °C/s when element is on (default 8)
  coolRate      number  — °C/s when element is off (default 3)
  maxTemp       number  — thermostat trip temperature in °C (default 100)
  resetTemp     number  — thermostat reset temperature in °C (default 70)
  minWatts      number  — standby draw always present, even when tripped (default 50)

  All power/capacitor/brownout logic is inherited from Load.
  States: inherited from Load ('off' | 'on' | 'brownout' | 'capacitor' | 'blown')
*/

class Heater extends Load {

    static type  = 'heater'
    static label = 'Heater'
    static group = 'Appliance'
    static dispatchDelay = 150  // faster than default — thermal events are time-sensitive

    static _defaultSpike() {
        return { enabled: true, percent: 30, duration: 3.8 }
    }

    static catalog = [
        { key: 'heater-1kw', label: 'Heater 1kW',  watts: 1000, minVolts: 200,
          spike: { enabled: true, percent: 30, duration: 0.8 } },
        { key: 'heater-2kw', label: 'Heater 2kW',  watts: 2000, minVolts: 205,
          spike: { enabled: true, percent: 35, duration: 0.9 } },
        { key: 'heater-3kw', label: 'Heater 3kW',  watts: 3000, minVolts: 210,
          spike: { enabled: true, percent: 40, duration: 1.0 } },
        { key: 'heater-oil', label: 'Oil Heater',   watts: 1500, minVolts: 200,
          spike: { enabled: true, percent: 15, duration: 0.5 } },
    ]

    static defaults(id, preset = {}) {
        return {
            ...super.defaults(id, preset),
            // Override label default from Load
            label:        preset.label    || 'Heater',
            watts:        preset.watts    ?? 1000,
            minVolts:     preset.minVolts ?? 200,
            // Thermal simulation
            temperature:  0,            // current temp in °C (0–maxTemp)
            heatState:    'cold',       // 'cold' | 'warming' | 'hot'
            heatSwitch:   true,         // thermostat: element enabled
            currentWatts: preset.minWatts ?? 50,  // live draw, starts at minimum
            heatRate:     preset.heatRate  ?? 8,  // °C/s while element is on
            coolRate:     preset.coolRate  ?? 3,  // °C/s while element is off
            maxTemp:      preset.maxTemp   ?? 100, // trip temperature
            resetTemp:    preset.resetTemp ?? 70,  // re-enable temperature
            minWatts:     preset.minWatts  ?? 50,  // standby draw (W)
            spike:        preset.spike ? { ...preset.spike } : { ...this._defaultSpike() },
        }
    }

    static configFields() {
        return [...super.configFields(), 'heatRate', 'coolRate', 'maxTemp', 'resetTemp', 'minWatts', 'spike']
    }

    // Override apply to use currentWatts (dynamic) instead of the fixed watts rating.
    static apply(panel, signal, graph) {
        if (panel.blown) return

        if (signal && signal.v > panel.maxVolts) {
            const prevState = panel.state
            panel.blown = true
            panel.state = 'blown'
            Heater.dispatch(panel, 'heater:blown', { volts: signal.v, maxVolts: panel.maxVolts })
            Heater.dispatch(panel, 'state:change', { from: prevState, to: 'blown' })
            graph.emit(panel, null)
            return
        }

        // We only need minWatts' worth of amps available to stay alive (standby).
        const minAmps  = panel.minWatts / NOMINAL_VOLTS
        const powered  = signal && signal.v >= panel.minVolts && signal.a >= minAmps

        const prevState = panel.state
        if (powered && prevState !== 'on' && prevState !== 'capacitor') NodeBase.startSpike(panel)
        const drawAmps = (panel.currentWatts / NOMINAL_VOLTS) * NodeBase.spikeMultiplier(panel)

        if (powered) {
            panel._lastGoodSignal = signal
            panel.state = 'on'
            graph.emit(panel, { v: signal.v, a: signal.a - drawAmps })
        } else if (!signal || signal.v <= 0) {
            if (panel.capacitance > 0 && panel.chargeWs > 0) {
                panel.state = 'capacitor'
                const held = panel._lastGoodSignal
                if (held) graph.emit(panel, { v: held.v, a: held.a - drawAmps })
                if (prevState !== 'capacitor')
                    Heater.dispatch(panel, 'heater:capacitor-failover', { chargeWs: panel.chargeWs })
            } else {
                panel.state = 'off'
                graph.emit(panel, null)
            }
        } else {
            panel.state = 'brownout'
            if (prevState !== 'brownout')
                Heater.dispatch(panel, 'heater:brownout', { volts: signal?.v, amps: signal?.a, minVolts: panel.minVolts })
            graph.emit(panel, null)
        }
        if (panel.state !== prevState)
            Heater.dispatch(panel, 'state:change', { from: prevState, to: panel.state })
    }

    static tick(panel, dt, graph) {
        // Capacitor logic (inherited)
        super.tick(panel, dt, graph)

        // Thermostat: trip element off at maxTemp, reset at resetTemp
        if (panel.heatSwitch && panel.temperature >= panel.maxTemp) {
            panel.heatSwitch = false
            Heater.dispatch(panel, 'thermostat:trip', { temp: panel.temperature })
        } else if (!panel.heatSwitch && panel.temperature <= panel.resetTemp) {
            panel.heatSwitch = true
            Heater.dispatch(panel, 'thermostat:reset', { temp: panel.temperature })
        }

        // Thermal update — only heat when element is on and the unit is powered
        const elementOn = panel.heatSwitch && (panel.state === 'on' || panel.state === 'capacitor')
        if (elementOn) {
            panel.temperature = Math.min(panel.maxTemp, panel.temperature + panel.heatRate * dt)
        } else {
            panel.temperature = Math.max(0, panel.temperature - panel.coolRate * dt)
        }

        const temp = +panel.temperature.toFixed(1)
        if (temp !== panel._lastCheckedTemp) {
            panel._lastCheckedTemp = temp
            Heater.throttle(panel, 'heater:temperature', { temp, max: panel.maxTemp })
        }

        // Dynamic draw: scale from minWatts (cold) up to rated watts (full temp)
        const heatFraction  = panel.temperature / panel.maxTemp
        const elementWatts  = panel.minWatts + (panel.watts - panel.minWatts) * heatFraction
        panel.currentWatts  = panel.heatSwitch ? elementWatts : panel.minWatts

        panel.heatState = panel.temperature < (panel.maxTemp * 0.2) ? 'cold'
                        : panel.temperature < (panel.maxTemp * 0.6) ? 'warming'
                        : 'hot'
        if (panel.heatState !== panel._lastHeatState) {
            Heater.dispatch(panel, 'heater:heat-state', { from: panel._lastHeatState, to: panel.heatState, temp: +panel.temperature.toFixed(1) })
            panel._lastHeatState = panel.heatState
        }

        // Re-apply with updated currentWatts so downstream sees the new amps
        // Only when the draw has meaningfully changed (>1 W) to avoid signal storms
        const prevWatts = panel._lastEmittedWatts ?? -1
        if (Math.abs(panel.currentWatts - prevWatts) > 1 &&
                panel.enabled !== false &&
                panel.signal !== undefined &&
                (panel.state === 'on' || panel.state === 'brownout')) {
            Heater.dispatch(panel, 'heater:draw-change', { from: +prevWatts.toFixed(1), to: +panel.currentWatts.toFixed(1) })
            panel._lastEmittedWatts = panel.currentWatts
            Heater.apply(panel, panel.signal, graph)
        }
    }

    static reset(panel, graph) {
        panel.temperature       = 0
        panel.heatState         = 'cold'
        panel._lastHeatState    = null
        panel._lastCheckedTemp  = null
        panel.heatSwitch        = true
        panel.currentWatts      = panel.minWatts ?? 50
        panel._lastEmittedWatts = -1
        Heater.dispatch(panel, 'heater:reset', {})
        super.reset(panel, graph)
    }
}

NodeRegistry.register(Heater)

File: power_graph/src/power_graph/nodes/heater.py
"""
nodes/heater.py
──────────────────────────────────────────────────────────────────────────────
Heater — extends Load with thermal simulation and thermostat.

Extra states: load 'on' / 'off' / 'brownout' / 'blown' states still apply,
but Heater adds a heatState field:

  heatState: 'cold' | 'warming' | 'hot'

Thermostat behaviour
  • temperature rises at heatRate °C/s while powered
  • temperature falls at coolRate °C/s while off / tripped
  • when temperature >= maxTemp  → thermostat cuts power (heatSwitch = False)
  • when temperature <= resetTemp → thermostat restores power (heatSwitch = True)

The effective wattage drawn modulates from minWatts → rated watts as the
element heats up (resistance effect).
"""

import math
from typing import Dict, List
from power_graph.node_base import NodeBase, Signal
from power_graph.node_registry import NodeRegistry
from .load import Load


class Heater(Load):
    """Resistive heater with thermal simulation."""

    type  = 'heater'
    label = 'Heater'
    group = 'Loads'
    dispatch_delay = 200

    catalog = [
        {'key': 'heater-1kw',  'label': 'Heater 1kW',     'watts': 1000, 'minWatts':  50, 'noise': 10, 'noiseInterval': 0.4, 'heatRate': 2.0, 'coolRate': 0.5, 'maxTemp': 80,  'resetTemp': 60},
        {'key': 'heater-2kw',  'label': 'Heater 2kW',     'watts': 2000, 'minWatts': 100, 'noise': 10, 'noiseInterval': 0.4, 'heatRate': 3.5, 'coolRate': 0.6, 'maxTemp': 80,  'resetTemp': 60},
        {'key': 'heater-3kw',  'label': 'Heater 3kW',     'watts': 3000, 'minWatts': 150, 'noise': 10, 'noiseInterval': 0.4, 'heatRate': 5.0, 'coolRate': 0.8, 'maxTemp': 80,  'resetTemp': 60},
        {'key': 'heater-oil',  'label': 'Oil Heater 2kW', 'watts': 2000, 'minWatts': 100, 'noise':  6, 'noiseInterval': 0.8, 'heatRate': 1.0, 'coolRate': 0.1, 'maxTemp': 75,  'resetTemp': 45},
    ]

    @classmethod
    def defaults(cls, node_id: int, preset: Dict = None) -> Dict:
        if preset is None:
            preset = {}
        base = super().defaults(node_id, preset)  # Load.defaults
        rated  = preset.get('watts', 1000)
        min_w  = preset.get('minWatts', rated * 0.05)   # ~5 % standby draw
        base.update({
            'temperature':         preset.get('temperature', 20.0),
            'heatState':           'cold',
            'heatSwitch':          True,               # thermostat relay
            'heatRate':            preset.get('heatRate',   2.0),   # °C/s
            'coolRate':            preset.get('coolRate',   0.5),   # °C/s
            'maxTemp':             preset.get('maxTemp',   80.0),
            'resetTemp':           preset.get('resetTemp', 60.0),
            'minWatts':            min_w,
            # currentWatts tracks the live scaled draw (minWatts → rated watts)
            # and is the value actually passed to Load.apply — panel['watts']
            # is NEVER modified so the rated ceiling is always preserved.
            'currentWatts':        min_w,
            '_last_emitted_watts': -1.0,
        })
        return base

    @classmethod
    def config_fields(cls) -> List[str]:
        return [*super().config_fields(), 'heatRate', 'coolRate', 'maxTemp', 'resetTemp']

    @classmethod
    def apply(cls, panel: Dict, signal: Signal, graph):
        rated          = panel['watts']
        min_w          = panel.get('minWatts', rated * 0.05)

        if not panel.get('heatSwitch'):
            # Thermostat tripped — element is off but standby draw remains.
            panel['watts'] = min_w
        else:
            panel['watts'] = panel.get('currentWatts', min_w)

        super().apply(panel, signal, graph)
        panel['watts'] = rated   # always restore the rated ceiling

    @classmethod
    def tick(cls, panel: Dict, dt: float, graph):
        super().tick(panel, dt, graph)  # Load.tick handles spike + noise + cap

        temp   = panel.get('temperature', 20.0)
        is_on  = panel.get('state') in ('on', 'brownout')

        if is_on and panel.get('heatSwitch'):
            temp += panel.get('heatRate', 2.0) * dt
        else:
            temp -= panel.get('coolRate', 0.5) * dt
            temp  = max(20.0, temp)   # ambient floor

        panel['temperature'] = round(temp, 2)

        # Thermostat transitions
        hs    = panel.get('heatSwitch', True)
        max_t = panel.get('maxTemp', 80)
        rst_t = panel.get('resetTemp', 60)

        if hs and temp >= max_t:
            panel['heatSwitch'] = False
            panel['heatState']  = 'hot'
            cls.dispatch(panel, 'heater:trip', {'temperature': temp})
        elif not hs and temp <= rst_t:
            panel['heatSwitch'] = True
            panel['heatState']  = 'warming'
            cls.dispatch(panel, 'heater:resume', {'temperature': temp})

        # heatState
        if temp < 30:
            panel['heatState'] = 'cold'
        elif panel.get('heatSwitch'):
            panel['heatState'] = 'warming'

        cls.throttle(panel, 'heater:temperature', {'temperature': panel['temperature'],
                                                    'heatState': panel['heatState']})

        # ── Dynamic draw: scale currentWatts from minWatts → rated watts ──────
        rated = panel['watts']           # never mutated — always the catalog value
        min_w = panel.get('minWatts', rated * 0.05)

        if panel.get('heatSwitch'):
            # Fraction is relative to the operating band (resetTemp → maxTemp) so
            # draw starts at minWatts when the thermostat re-enables and only
            # reaches rated watts at maxTemp — not at 75% on the very first tick.
            band      = max(1.0, max_t - rst_t)
            heat_frac = min(1.0, max(0.0, (temp - rst_t) / band))
            target    = min_w + (rated - min_w) * heat_frac
        else:
            target = min_w

        # Apply noise ripple directly to the thermal-scaled draw.
        # Load.tick() (called above via super()) already advanced _noise_phase,
        # so we just read it here — no extra state needed.
        noise_cfg = panel.get('noise', 0)
        if isinstance(noise_cfg, dict):
            noise_pct = noise_cfg.get('amount', 0) if noise_cfg.get('enabled') else 0
        else:
            noise_pct = noise_cfg
        if noise_pct and panel.get('state') == 'on':
            offset = math.sin(panel.get('_noise_phase', 0.0)) * noise_pct / 100
            target = max(min_w, target * (1.0 + offset))

        panel['currentWatts'] = round(target, 1)

        # Always keep current_watts in sync — the generator BFS reads this field
        # directly to calculate drawWatts.  Without this, Load.tick() (called via
        # super() above) overwrites current_watts with rated * noise and the
        # generator sees full draw even while the thermostat is tripped.
        panel['current_watts'] = panel['currentWatts']

        # Re-apply whenever the draw shifts (lower threshold so every noisy
        # tick propagates upstream and the generator sees the ripple).
        prev = panel.get('_last_emitted_watts', -1.0)
        if (abs(panel['currentWatts'] - prev) > 0.5
                and panel['state'] in ('on', 'brownout')
                and panel.get('_last_signal') is not None):
            panel['_last_emitted_watts'] = panel['currentWatts']
            cls.apply(panel, panel['_last_signal'], graph)


NodeRegistry.register(Heater)
